File: server/src/main/java/me/ajfleming/tworoomsio/engine/UserManager.java
package me.ajfleming.tworoomsio.engine;

import com.corundumstudio.socketio.SocketIOClient;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import me.ajfleming.tworoomsio.exception.UserException;
import me.ajfleming.tworoomsio.model.User;
import me.ajfleming.tworoomsio.socket.event.ReloadGameSessionEvent;
import org.springframework.stereotype.Component;

@Component
public class UserManager {

  /**
   * Mapped as Key = Socket ID
   */
  Map<String, User> connectedUsers;
  /**
   * Mapped as Key = User Token
   */
  Map<String, User> disconnectedUsers;

  public UserManager() {
    connectedUsers = new HashMap<>();
    disconnectedUsers = new HashMap<>();
  }

  public User createUser(SocketIOClient client, String name) {
    User user = new User(name, client);
    connectedUsers.put(user.getSocketSessionId().toString(), user);
    return user;
  }

  public Optional<User> getUser(SocketIOClient client) {
    return Optional.ofNullable(connectedUsers.get(sessionId(client)));
  }

  public Optional<User> reconnectDisconnectedUser(SocketIOClient client,
      ReloadGameSessionEvent event) {
    User user = disconnectedUsers.get(event.getPlayerToken());
    if (user == null) {
      Optional<User> connectedUserResult = findConnectedUserByPlayerId(event.getPlayerToken());
      if (connectedUserResult.isPresent()) {
        user = connectedUserResult.get();
        user.getClient().disconnect();
        user.setClient(client);
      }
    }
    if (user != null && user.authenticateUser(event.getPlayerToken(),
        event.getPlayerSecret())) {
      user.setClient(client);
      disconnectedUsers.remove(event.getPlayerToken());
      connectedUsers.put(client.getSessionId().toString(), user);
      return Optional.of(user);
    } else {
      return Optional.empty();
    }
  }

  public Optional<User> handleDisconnectedClient(SocketIOClient client) {
    User user = connectedUsers.get(sessionId(client));
    if (user != null) {
      user.disconnectPlayer();
      disconnectedUsers.put(user.getUserToken(), user);
      return Optional.of(user);
    }
    return Optional.empty();
  }

  public void sendEvent(String playerToken, String eventName, Object payload) throws
      UserException {
    Optional<User> userResult = findConnectedUserByPlayerId(playerToken);
    if (userResult.isPresent()) {
      userResult.get().sendEvent(eventName, payload);
    } else {
      throw new UserException("Failed to send event to user");
    }
  }

  private Optional<User> findConnectedUserByPlayerId(String playerToken) {
    return connectedUsers.values().stream().filter(user -> user.getUserToken().equals(playerToken))
        .findFirst();
  }

  private String sessionId(SocketIOClient client) {
    return client.getSessionId().toString();
  }
}

File: kotlin_server/src/main/kotlin/me/ajfleming/tworoomsio/engine/UserManager.kt
package me.ajfleming.tworoomsio.engine

import com.corundumstudio.socketio.SocketIOClient
import me.ajfleming.tworoomsio.model.Player
import org.springframework.stereotype.Component
import kotlin.collections.HashMap

@Component
class UserManager(
    private var connectedUsers: HashMap<String, Player> = HashMap(),
    private var disconnectedUsers: HashMap<String, Player> = HashMap()
) {
    fun createUser(client: SocketIOClient, name: String): Player {
        val user = Player(name, client)
        connectedUsers[user.userToken] = user
        return user;
    }

    fun getUser(client: SocketIOClient) : Player? {
        return connectedUsers[sessionId(client)]
    }

    fun reconnect(client: SocketIOClient, userToken: String, userSecret: String) : Player? {
        var user = disconnectedUsers[userToken]
        if (user == null) {
            user = findConnectedUserByUserToken(userToken)
            if (user != null) {
               user.client.disconnect()
            }
        }
        if(user != null && user.authenticate(userToken, userSecret)) {
            user.client = client
            disconnectedUsers.remove(userToken)
            connectedUsers[userToken] = user
        }
        return user
    }

    fun handleDisconnect(client: SocketIOClient) : Player? {
        val user = connectedUsers[sessionId(client)]
        if (user != null) {
            user.disconnectPlayer()
            disconnectedUsers[user.userToken] = user
        }
        return user
    }

    fun findConnectedUserByUserToken(userToken: String) : Player? {
        return connectedUsers[userToken]
    }

    private fun sessionId(client: SocketIOClient) : String {
        return client.sessionId.toString()
    }
}